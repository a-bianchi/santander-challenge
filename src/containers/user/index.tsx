import { ReactElement } from 'react';
const User = (): ReactElement => {
    return (
        <div>
            <p>Soy User</p>
        </div>
    );
};

export default User;
