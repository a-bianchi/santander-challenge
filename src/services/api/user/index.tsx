import * as Types from './user.types';
import { User } from './user.types';

export class UserApi {
    private users = [
        {
            id: '1',
            username: 'Admin',
            password: 'Admin',
            role: 'Admin',
            age: 21,
        },
        {
            id: '2',
            username: 'User',
            password: 'User',
            role: 'User',
            age: 21,
        },
        {
            id: '3',
            username: 'User2',
            password: 'User2',
            role: 'User',
            age: 17,
        },
    ];

    /**
     * Login.
     */
    async login(username: string, password: string): Promise<Types.GetUserResult> {
        // make the api call

        const user = this.users.filter((user) => user.username === username && user.password === password);

        const response = user[0]
            ? {
                  ok: true,
                  problem: null,
                  originalError: null,
                  data: user[0],
                  status: 200,
              }
            : {
                  ok: false,
                  problem: 'CLIENT_ERROR',
                  originalError: {
                      name: 'error',
                      code: '401',
                      config: { url: 'null' },
                      isAxiosError: true,
                      toJSON: {},
                  },
                  data: {
                      id: '',
                      username: '',
                      password: '',
                      role: 'User',
                      age: '',
                  },
                  status: 401,
              };

        // the typical ways to die when calling an api
        if (!response.ok) {
            return {
                kind: 'unauthorized',
            };
        }

        // eslint-disable-next-line @typescript-eslint/no-explicit-any
        const convert = (raw: any): User => {
            return {
                id: raw?.id,
                username: raw?.username,
                password: '',
                role: raw?.role,
                age: raw?.age,
            };
        };

        // transform the data into the format we are expecting
        try {
            return { kind: 'ok', user: convert(response.data) };
        } catch {
            return { kind: 'bad-data' };
        }
    }
}
