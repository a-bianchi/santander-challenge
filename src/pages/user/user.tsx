import { ReactElement, useEffect, useState } from 'react';
import { observer } from 'mobx-react-lite';
import { useHistory } from 'react-router-dom';
import { isAdmin } from '../../utils/storage';
import { NavBar, DataTable, DataTableType } from '../../components';
import { Container } from '@material-ui/core';
import { useTranslation } from 'react-i18next';
import { useStores } from '../../models';
import { MeetupApi } from '../../services/api/api';

const mockData = [
    {
        title: 'Meetup mock',
        date: '20/02/2021',
        temperature: 23,
        people: 66,
        beers: 66,
        status: 'confirmed',
    },
];

const convert = (row): DataTableType => {
    return {
        title: row.title,
        date: row.data,
        temperature: row.temperature,
        people: row.people,
        beers: row.beers,
        status: row.status,
    };
};

export const User = observer(
    (): ReactElement => {
        const history = useHistory();
        const { t } = useTranslation();
        const [meetups, setMeetups] = useState(mockData);
        const meetup = new MeetupApi();
        const { userStore } = useStores();

        const actionCheckIn = [
            {
                name: t('user.checkIn'),
                handler: () => {
                    console.log('check in');
                },
            },
        ];

        // const actionGoto = [
        //     {
        //         name: t('user.Goto'),
        //         handler: () => {
        //             console.log('go to');
        //         },
        //     },
        // ];

        useEffect(() => {
            if (isAdmin()) {
                history.push('/admin');
            }
            const userId = userStore.id || '2';
            meetup
                .getUserAttendMeetup(userId)
                .then((data) => {
                    if (data && data.kind === 'ok') {
                        const arrayMeetups = data.meetups.map((meet) => convert(meet));
                        setMeetups(arrayMeetups);
                    }
                })
                .catch((error) => {
                    console.log(`Error: getUserMeetup ${error}`);
                });
        }, []);

        return (
            <div>
                <NavBar data-test="NavBar" admin={false} />
                <Container maxWidth="md">
                    <DataTable
                        data-test="DataTable"
                        title={t('user.tableTitle')}
                        actions={actionCheckIn}
                        data={meetups}
                    />
                </Container>
            </div>
        );
    },
);
