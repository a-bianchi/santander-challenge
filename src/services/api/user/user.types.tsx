import { GeneralApiProblem } from '../api.problem';

export interface User {
    id: string;
    username: string;
    password: string;
    role: 'Admin' | 'User';
    age: number;
    token: string;
}

export interface Invite {
    status: string;
    message: string;
}

export type GetUserResult = { kind: 'ok'; user: User } | GeneralApiProblem;
export type InviteUserResult = { kind: 'ok'; invite: Invite } | GeneralApiProblem;
