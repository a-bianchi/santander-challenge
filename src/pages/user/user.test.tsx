/* global describe expect it */
import { shallow } from 'enzyme';
import { User } from './user';

describe('User Page', () => {
    it('should be defined', () => {
        expect(User).toBeDefined();
    });
    it('should render correctly', () => {
        const tree = shallow(<User />);
        expect(tree.find('.NavBar')).toBeDefined();
        expect(tree.find('.DataTable')).toBeDefined();
        expect(tree).toMatchSnapshot();
    });
});
