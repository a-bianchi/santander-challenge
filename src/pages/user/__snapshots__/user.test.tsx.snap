// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`User Page should render correctly 1`] = `ShallowWrapper {}`;
